### DISCUSSION TITLE
Storybook ‘Show Code’ issue

### DISCUSSION THREAD

**Discussion by ShraddhaGurav** [storybookjs/storybook#25135] [Help]
### Summary

Hello,

Currently I am working on a storybook project build with angular framework. I have created a custom library with basic components.
When I am using the library component within another storybook component, Show Code does not show the complete code of the component reused from library, instead only shows the selector of the storybook component.
For example: 

lib-component is a package and has custom button module, this button module is used within storybook button component. 

Output shown: 
![image](https://github.com/storybookjs/storybook/assets/13862064/7265c048-5cda-4b54-8171-a2a57252f6fc)

expected output should have <button> in place of <dafm-button-test>

**Comment by Marklb** [comment]
Can you provide a reproduction or maybe the story? I am not understanding what you are asking.

Are you saying that you have something, like the following?
```ts
@Component({
  selector: 'lib-component',
  template: ``,
  standalone: true,
})
class LibComponent {}

@Component({
  selector: 'dafm-button-test',
  template: `<lib-component></lib-component>`,
  imports: [ LibComponent ],
  standalone: true,
})
class DAFMComponent {
  @Input() label: string
  @Input() disabled: boolean
  @Input() class: string
  @Output() onClick = new EventEmitter<ClickEvent>()
}

export default {
  component: DAFMComponent
}

const Story1 = {
  args: {
    label: 'Primary button',
    disabled: false,
    class: 'primary-button',
  },
  argTypes: {
    onClick: { action: 'clicked' },
  },
}

const Story2 = {
  render: args => ({
    props: args,
    template: `<dafm-button-test
        [label]="'Primary button'"
        [disabled]="false"
        [class]="'primary-button'"
        (onClick)="onClick($event)">
      </dafm-button-test>
      <lib-component></lib-component>`
  }),
  decorators: [
    moduleMetadata({ imports: [ LibComponent ] }),
  ],
  argTypes: {
    onClick: { action: 'clicked' },
  },
}
```
If one of those is similar to your situation, `Story1` would show your screenshot and `Story2` would show your screenshot with "<lib-component></lib-component>". It shows the story template, not what is currently rendered in the DOM.

**Reply by ShraddhaGurav** [reply]
Hello,

For example:

This is the button library:

@Component({
  selector: 'app-button',
  template: `
<button mat-flat-button 
    [ngClass]="class"
    (click)="onClick.emit($event)"
    [disabled]="disabled">
    <span class="button-label-text">{{ label }} </span>
</button>
`,
  styleUrls: ['./button.component.scss']
})
export class ButtonComponentLib {
  @Input() label?: string;
  @Input() disabled: boolean = false;
  @Input() class?: string;
  @Output() onClick = new EventEmitter<Event>();
}

I am importing the above library in a storybook button component as follows:

@Component({
  selector: 'button-test',
  standalone: true,
  imports: [CommonModule, MatButtonModule, ButtonComponentLib],
  template: `
<app-button
  [label]="label"
  [disabled]="disabled"
  [class]="class"
  (onClick)="onClick.emit($event)"></app-button>
`,
  styleUrls: ['./button.scss'],
})

export default meta;
type Story = StoryObj<Button>;

export const Primary: Story = {
  args: {
    label: 'Primary button',
    disabled: false,
    class: 'primary-button',
  }
};

When I check 'Show Code' it just displays,

`<button-test></button-test>`

while the expected output should be, 

```
<app-button
  [label]="'Primary button'"
  [disabled]="false"
  [class]="'primary-button'"
  (onClick)="onClick($event)"
></app-button>
```

**Reply by joevaugh4n** [reply]
Thanks for jumping in here, @Marklb!

**Reply by Marklb** [reply]
@ShraddhaGurav I still don't know what exactly you are doing, from the partial snippets. Without the `Meta` of your story, I can only see the args, so I don't know what component the story is for. *I assume `Meta` has `component: ButtonTestComponent`, based on the output.*

Does this happen for all stories? Do you have compodoc enabled?

I tried to create a reproduction for you, but I had to mostly guess. I included a way to add a story that causes the problem, by returning a `template` in the story `render` function, and how to fix it, but you didn't mention doing that, so I don't know if it is relevant. If the following Stackblitz doesn't show what you are experiencing, could you fork it and change it to what you are doing? https://stackblitz.com/edit/github-6qncyk?file=src%2Fapp%2Fexample%2Fbutton.stories.ts,src%2Fapp%2Fexample%2Ftest-button.stories.ts

**Reply by ShraddhaGurav** [reply]
@Marklb Thank you for the propmt replies. I have modified the Stackblitz: https://stackblitz.com/edit/github-6qncyk-hg6rsr?file=src%2Fapp%2Fexample%2Ftest-button.component.ts

Here I am importing **ButtonComponentLib** module in **TestButtonComponent**. 
ButtonComponentLib is an independent angular component, whereas TestButtonComponent is a storybook component.

The output that we are expecting in Show Code is 
```
<app-button
  [label]="'Primary button'"
  [disabled]="false"
  [class]="'primary-button'"
  (onClick)="onClick($event)"
></app-button>
``` 
instead of 

```
<test-button
  [label]="'Primary button'"
  [disabled]="false"
  [class]="'primary-button'"
  (onClick)="onClick($event)"
></test-button>
```

The project requirement is to display module/library component code instead of storybook component code in Show Code, if the particular module is imported in storybook component. Is there any way which can be used to manipulate the default behaviour?

**Reply by Marklb** [reply]
@ShraddhaGurav I think I understand now and I think there is a misunderstanding with how Storybook uses components. There isn't anything special that would make a component a "Story Component" in the sense that I think you are referring to. As far as Storybook is concerned, `TestButtonComponent` is just an Angular component, like `ButtonComponentLib`.

By doing the following:
```ts
const meta: Meta<TestButtonComponent> = {
  component: TestButtonComponent,
  tags: ['autodocs'],
};

export default meta;
```
That tells Storybook that this file contains stories for `TestButtonComponent`, so anything automatically generated is going to be under the assumption that it is for using `TestButtonComponent`. By doing that, you are just adding an additional component to maintain that doesn't have access to the exact type or comments from your component, unless you manually duplicate everything in what you were calling a "Story Component".

Ideally, you should be able to just add a component to a stories file `Meta` and simply add stories that are just exported objects with different args that get used as inputs. Of course, they can't all be that simple, but the docs have examples for different ways to provide your own template, if the one generated for you by Storybook is not able to express yours full story, if you need to add additional imports /declarations/providers, etc. https://storybook.js.org/docs/writing-stories
